CONTRIBUTING.md

```diff
@@ -1,57 +1,77 @@
 
-Again, this makes review much easier.
+Features should always include unit tests or integration tests where possible.
 
-Please try not to submit pull requests that don't add new features (e.g. moving stuff around)
-unless you see something that is obviously wrong or that could be written in a more terse or
-idiomatic style. It takes time to review each pull request - time that I'd prefer to spend writing
-new features!
+> One exception to this is features related to DevTools which has no tests currently
 
-Prefer terseness to verbosity but don't try to be too clever.
+Features that introduce new controls should consider the following:
 
-## Tests
+- Ideally the control should be exposed to the operating system's automation/accessibility APIs by writing an `AutomationPeer`
+- If the control introduces any functionality which is difficult to unit test, an integration test should be written
 
-There are two types of tests currently in the codebase; unit tests and render tests.
+## General Guidance
 
-Unit tests should be contained in a class name that mirrors the class being tested with the suffix
--Tests, e.g.
+## PR description
 
-    Avalonia.Controls.UnitTests.Presenters.TextPresenterTests
+- The PR template contains sections to fill in. These are discretionary and are intended to provide guidance rather than being prescritive: feel free to delete sections that do not apply, or add additional sections
+- **Please** provide a good description of the PR. Not doing so **will** delay review of the PR at a minimum, or may cause it to be closed. If English isn't your first language, consider using ChatGPT or another tool to write the description
+- Link any fixed issues with a `Fixes #1234` comment
 
-Where Avalonia.Controls.UnitTests is the name of the project.
+## Breaking changes
 
-Unit test methods should be named in a sentence style, separated by underscores, that describes in
-English what the test is testing, e.g.
+- During a major release cycle, source or binary breaking changes may not be introduced to the codebase: this is checked by an automated tool and will cause CI to fail
+- If something needs addressing in the next major release, you can leave a `TODOXX:` comment, where `XX` is the version number of the next major release, e.g. `TODO12:`
+- Carefully consider behavioral breaking changes and point them out in the PR description
+
+### Commits
+
+In addition to the guidance in the `Bug Fixes` section, please follow these guidelines:
+
+- Rebase your changes to remove extraneous commits. Ideally the commit history should tell a clean story of how the PR was implemented (even though the process was probably not clean!)
+- Provide meaningful commit comments
+- **Do not** change code unrelated to the bug fix/feature
+- **Do not** introduce spurious formatting or whitespace changes
```

Comment:
I know we have differences of option here but I think this is too strict of a policy. It doesn't allow fixes to ever occur unless the entire file is replace. Code churn as you put it needs to happen.

IMO we need:
 1. An actual code formatting guide with enforced rules in an .editorconfig (has been needed for a while)
 2. Relax formatting changes like this so they are allowed if you notice something in the file you are already editing. Most projects allow fixes if you are already working in the file or its related to the feature you are working on.
 3. Clarify that formatting changes should only be of clear benefit: Modern syntax like `get { return ...` to `get => ...`, remove blank lines with spaces, add space after // in comment, etc. These are unarguable for the most part. New devs will sometimes open PRs changing things that are unnecessary but that can be discussed case-by-case. I don't think it should be prohibited outright.

Finally, I get the impression you are a purist with git history. Maintaining a clear and easy to read history is most important to you. I've found its often better to just let the code evolve. It's rare that the full git history needs to be investigated. And when it does need to be investigated it can still be done even with formatting changes. It just requires a little more work. Bottom line I rank the cost/benefit different than you and prioritize code quality over git history quality.

Reply:
It's all logical and stuff, but you aren't the one who has to go through the file history when diagnosing issues because random change has polluted `git blame`. We are doing it quite often and it's a rather annoying procedure.
And unlike preparing a PR it will have to be done _forever_ while the code exists.

Another problem is reverting PRs that caused regressions. If PR was merged several months ago and it had formatting changes, there is a huge chance that it won't get reverted cleanly because the affected random pieces of code were already changed by other PRs. e. g. it was pretty much impossible to revert a change made a while ago which was rather frustrating and have cost me several hours of my life. 

So I'd rather _**ignore the PR completely**_ than increase the maintenance burden because of somebody's ideas of beauty.

Reply:
TL;DR: PR author is not the one who will have to maintain the code ad aeternam, so clean history is more important.

Reply:
> because of somebody's ideas of beauty.

It's not beauty, it's code quality I'm advocating for. I think code quality is much more important than perfect git history. In fact I've never heard of someone who will refuse fixes because it will make the git history more complicated.

Again, I propose a few compromises;
 - Only focus on clear improvements (which can be agreed ahead of time with an editorconfig rule set)
 - Only allow formatting fixes for files already being edited for other reasons

You could also still allow formatting-only PRs that would fix stuff like this.

Several make commits without detailed review and a lot of formatting mistakes get made. I often see missed spaces after `if` for example in addition to what I stated above. It doesn't make sense not to allow fixes this. If you stand firm that these shouldn't be fixed I'll argue that we need to be much, much more careful during PR merges. In fact we should be running code formatting checks before allowing merge if you want to go this route.

> maintain the code ad aeternam, so clean history is more important.

That logic doesn't seem to hold. The main reasons to go through the history are to find the original authors and ask them to take a look. Or to find the related PR's, issues and discussions to understand more background. As you already said, original authors come and go so there is a decreasing significance over time with git blame (significance approaches zero over project lifetime).

Concerning related PR's issues and discussions several projects provide a lot of comments and references in code to avoid having to dig through to find related issues and discussions. In the end PR is still useful to get to with git history though.

Reply:
> but you aren't the one who has to go through the file history when diagnosing issues because some random change has polluted git blame. We are doing it quite often and it's a rather annoying procedure.

Also, are you using bit bisect for this? It can filter past a lot of commits quickly.

Reply:
These guidelines are aimed at improving the speed in which we can review PRs. You may have noticed that we have an ever-growing list of open PRs and not enough time to review them.

Not following these guidelines doesn't mean we close your PR, but it does mean that it's less likely to get reviewed in a timely manner. Personally, my brain gets _very_ distracted by unrelated changes so much that I'll move onto some other urgent job instead of finishing reviewing.

Reply:
> Also, are you using git bisect for this? It can filter past a lot of commits quickly.

Yes, i use this almost daily and non-building or intermediate commits make it a terrible experience.